### ISSUE TITLE
Shortcode tags in automatically generated descriptions

### ISSUE THREAD

**Issue by chesio** [gnotaras/wordpress-add-meta-tags#9]
Hi George,

I'd like to resolve one of TODO items in your code ;) More precisely, I'd like to follow up the issue we already discussed at WP forums: https://wordpress.org/support/topic/shortcode-tags-in-automatically-generated-descriptions

I found out that function `amt_get_the_excerpt` still can produce description with shortcode in it if generated excerpt (content of `$amt_excerpt` variable) is shorter than average description length and ["nasty hack"](https://github.com/gnotaras/wordpress-add-meta-tags/blob/master/amt-utils.php#L300) kicks in. The problem is that "nasty hack" produces **new excerpt from raw post content** - such excerpt is processed via `amt_sanitize_description` [function](https://github.com/gnotaras/wordpress-add-meta-tags/blob/master/amt-utils.php#L123), but as this function only cleans double quotes by default, shortcode is not removed.

However, you seem to [be aware that it might be not enough](https://github.com/gnotaras/wordpress-add-meta-tags/blob/master/amt-utils.php#L129). Well, now you know the answer is negative :) At least unless you re-implement or remove "nasty hack".

Btw. thanks for investigating the issue while on WP forums!

All the best,
Česlav

**Comment by gnotaras** [comment]
Hi Česlav,

Thanks for your feedback. Indeed, it seems there is an issue in there. I'll try to check it thoroughly in the following days.

Best Regards,
George

**Comment by chesio** [comment]
Hi George,

Thanks for your work. It's not a big issue, I simply fixed it by commenting out the "nasty hack". I'd suggest you to drop it completely - I cannot think of any situation in which it could help to produce better excerpt. Obviously it can produce longer excerpt, but only by including some of the garbage that has been removed from post content by removal of HTML tags and shortcode.

Best regards,
Česlav

**Comment by gnotaras** [comment]
I see. I didn't have enough time to test this. Indeed, this was needed in the long past and haven't really tested if it is still required.

I suppose I'll add a filter-based switch to short-circuit it. That will be a first step. It will be possibly removed in completely in future releases.

Could you please let me know exactly of the part of the code you would like to exclude? Is it just the following part?

```php
        // If this was not enough, try to get some more clean data for the description (nasty hack)
        if ( strlen($amt_excerpt) < $desc_avg_length ) {
            $amt_excerpt = sanitize_text_field( amt_sanitize_description( substr($post->post_content, 0, (int) ($excerpt_max_len * 1.5)) ) );
            if ( strlen($amt_excerpt) < $desc_avg_length ) {
                $amt_excerpt = sanitize_text_field( amt_sanitize_description( substr($post->post_content, 0, (int) ($excerpt_max_len * 2)) ) );
            }
        }
```

George

**Comment by gnotaras** [comment]
> The problem is that "nasty hack" produces new excerpt from raw post content

Česlav, I'm really I hadn't realized this earlier. This is a horrible bug. Thanks for pointing it out. The nasty hack should operate on the clean post content. Thanks for pointing it out! It will be fixed ASAP.

**Comment by gnotaras** [comment]
I'd be grateful if you tested the following modification and let me know if it fixes the issue without having to remove the the code:

```php
        // If this was not enough, try to get some more clean data for the description (nasty hack)
        if ( strlen($amt_excerpt) < $desc_avg_length ) {
            $amt_excerpt = sanitize_text_field( amt_sanitize_description( substr($plain_text_processed, 0, (int) ($excerpt_max_len * 1.5)) ) );
            if ( strlen($amt_excerpt) < $desc_avg_length ) {
                $amt_excerpt = sanitize_text_field( amt_sanitize_description( substr($plain_text_processed, 0, (int) ($excerpt_max_len * 2)) ) );
            }
        }
```

Basically `$post->post_content` has been replaced with the cleaned `$plain_text_processed`.

Thanks in advance.

**Comment by chesio** [comment]
Hi George,

Yep, the fix helps, thanks!

**Comment by gnotaras** [comment]
Thanks for letting me know. This will be fixed in the next release. Big kudos for noticing this!

**Comment by gnotaras** [comment]
This has been fixed in v2.9.6. Thank you for your feedback. Much appreciated!

I will now close this issue.
